docs/proposals/provider-agnostic-dns-health-checks.md

```diff
@@ -0,0 +1,101 @@
+This feature has a few shortfalls:
+
+1. It’s tightly coupled with Route53. If other DNS providers are supported they must either provide a similar feature, or health checks will not be supported
+2. Lacks the ability to reach endpoints in private networks
+
+This document describes a proposal to extend the current health check implementation to overcome these shortfalls.
+
+### Goals
+
+* Ability to configure health checks in the DNSPolicy associated to a Gateway
+* Ability for the health checks to reach endpoints in private networks
+* DNS records are disabled when the associated health check fails
+* Current status of the defined health checks is visible to the end user
+* Transparently keep support for other health check providers like Route53
+
+## Proposal
+
+### Control plane
+
+#### `DNSHealthCheck` resource
+
+The DNSHealthCheck resource represents the configuration of a health check. The reconciliation of this resource results in the creation of health checks in the selected DNS provider, or, if configured to use the agent, the creation of a health check probe in the downstream cluster
+
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSHealthCheck
+metadata:
+  name: example-dnshealthcheck
+  ownerReference:
+    type: DNSPolicy
+    name: example-policy
+spec:  
+  config:
+    port: "..."
+    path: "..."
+  provider: AWS | Agent | ...
+  providerSpecific:
+    "agent/host": "..."
+status:
+  providerSpecific:
+    "agent/health-status": "Healthy"
+    "agent/last-check": "..."
+    "aws/health-check-id": "..."
+```
+
+Relevant fields
+
+| Field name | Description |
+| ---------- | ----------- |
+| `metadata.ownerReference` | Reference to the owner of the health check. Keeps track in order to propagate deletion and reconciliation. An update to the DNSHealthCheck should trigger the reconciliation of the owner resource |
+| `spec.config` | Provider agnostic configuration section of the health check |
+| `spec.provider` | Name of the health check provider to use |
+| `spec.providerSpecific` | Provider specific configuration |
+| `status.providerSpecific` | Provider specific status |
+
+The reconciliation logic of this resource broadly follows:
+
+1. Select the health check reconciler implementation based on the value of the provider field
+1. Delegate the reconciliation on the health check reconciler, passing the configuration and the provider specific values
+1. Update the status and the providerSpecific status values from the reconcile result
+1. The provider specific status values are propagated to the related DNSRecord
+
+### Data plane
+
+#### `DNSHealthCheckProbe` resource
```

Comment:
Something I haven't decided yet is whether to make the agent a controller that watches the DNSHealthCheckProbe resource, or a separate system that is managed by the DNSHealthCheckProbe controller

Reply:
I would expect it should be a controller reconciling DNSHealthCheckProbes, but might be missing something?

Reply:
I think a controller that reconciles the resource and pushes the checks into a queue and sets up a worker to process that check every n seconds. If the check changes the queue is flushed and the worker recreated to perform the check based on the new time WDYT